README.md
# ソケット (soketto) server

TBA

## Documentation

[API Docs](docs/README.md)


CHANGELOG.md
# Semantic Release Changelog

## [1.0.3](https://github.com/tufcode/soketto-server/compare/v1.0.2...v1.0.3) (2023-04-20)


### Performance Improvements

* removed ping & error ([9c37281](https://github.com/tufcode/soketto-server/commit/9c372813a675f93565ea959f2d21ddac9c33223d))

## [1.0.2](https://github.com/tufcode/soketto-server/compare/v1.0.1...v1.0.2) (2023-04-20)


### Performance Improvements

* upgrade paketto version ([80feb11](https://github.com/tufcode/soketto-server/commit/80feb11e29ab3f8b9d1e7b3bd79d2207a4756b43))

## [1.0.1](https://github.com/tufcode/soketto-server/compare/v1.0.0...v1.0.1) (2023-04-19)


### Bug Fixes

* things work now, hopefully ([7b65f9e](https://github.com/tufcode/soketto-server/commit/7b65f9ef26f42cb000c473463aa7732bffea8ba3))

# 1.0.0 (2023-04-18)


### Bug Fixes

* **ci:** Move generated docs to repository ([b302cc3](https://github.com/tufcode/soketto-server/commit/b302cc3c0228bbd907ea122240e8197f13748823))

## [1.0.1](https://github.com/daisyengine/server/compare/v1.0.0...v1.0.1) (2023-04-06)


### Bug Fixes

* **publish:** Almost forgot ([70db56e](https://github.com/daisyengine/server/commit/70db56e7275742837123987f1a0bb9f1d34d1f63))


src/SocketClient.ts
import { serializeUInt8 } from "paketto";
import { ClientStatus } from "./Enums/ClientStatus";
import { Protocol } from "./Enums/Protocol";
import { ITransport } from "./Transports/ITransport";
import { Server } from "./Server";

export class SocketClient {
  /**
   * The id of this client
   * @internal
   */
  private readonly _id: number;

  /**
   * The transport this client is using
   * @internal
   */
  private readonly _transport: ITransport;

  private _server: Server;

  get id() {
    return this._id;
  }

  get transport() {
    return this._transport;
  }

  /**
   * Status of this client. Do not modify this value.
   * @internal
   */
  status: ClientStatus = ClientStatus.CONNECTED;

  /**
   * Custom data associated with this client. You may set this to
   * anything.
   */
  data: unknown;

  onMessage: (event: string | number, data: unknown) => void = () => {};
  onClose: () => void = () => {};
  onError: () => void = () => {};

  /**
   * @internal
   */
  constructor(id: number, transport: ITransport, server: Server) {
    this._id = id;
    this._transport = transport;
    this._server = server;
  }

  /**
   * Disconnects this client
   */
  close() {
    if (this.status >= ClientStatus.CLOSING) return;
    this.status = ClientStatus.CLOSING;

    this._transport.close(this._id);
  }

  /**
   * Sends a message to this client
   * @param event Unique identifier for this event.
   *
   * It is best to use a string or number for this, but you can use
   * any type that can be serialized. Recommended type is uint8 (any number
   * between 0 and 255).
   *
   * @param data Data that will be sent to this client.
   */
  send(event: unknown, data: unknown) {
    if (this.status >= ClientStatus.CLOSING) return;

    let offset = serializeUInt8(this._server.buffer, Protocol.SC_Message);
    offset = this._server.paketto.serialize(this._server.buffer, event, offset);
    offset = this._server.paketto.serialize(this._server.buffer, data, offset);

    this._transport.send(this._id, this._server.buffer.subarray(0, offset));
  }
}


src/Transports/ITransport.ts
import { Server } from "../Server";

export interface ITransport {
  /**
   * The server this transport is attached to. This is set by the server
   * when the transport is added.
   */
  server: Server;

  /**
   * Should be called when a new connection is established.
   * @param connectionId The ID of the connection.
   */
  onConnect: (connectionId: number) => void;
  /**
   * Should be called when a connection is closed.
   * @param connectionId The ID of the connection.
   */
  onDisconnect: (connectionId: number) => void;
  /**
   * Should be called when a message is received.
   * @param connectionId The ID of the connection.
   * @param message The message received.
   */
  onMessage: (connectionId: number, message: Uint8Array) => void;

  /**
   * Sends a message to a connection.
   * @param connectionId The ID of the connection.
   * @param message The message to send.
   */
  send(connectionId: number, message: Uint8Array): void;
  /**
   * Closes a connection.
   * @param connectionId The ID of the connection.
   */
  close(connectionId: number): void;

  /**
   * Starts the transport.
   */
  start(): void;
  /**
   * Stops the transport.
   */
  stop(): void;

  /**
   * Gets the IDs of all clients connected to this transport.
   */
  getConnections(): number[];
}


src/Transports/WebSocket/WebSocketTransport.ts
import * as uWS from "uWebSockets.js";
import { ITransport } from "../ITransport";
import { Server } from "../../Server";

export class WebSocketTransport implements ITransport {
  server: Server;

  /**
   * @internal
   */
  private _app: uWS.TemplatedApp;
  /**
   * @internal
   */
  private _clients: Map<number, uWS.WebSocket>;
  /**
   * @internal
   */
  private _ids: Map<uWS.WebSocket, number>;
  /**
   * @internal
   */
  private _listen_socket: uWS.us_listen_socket;
  /**
   * @internal
   */
  private _port = 3000;

  get port() {
    return this._port;
  }

  set port(port: number) {
    if (this._listen_socket) {
      throw new Error("Cannot change port while listening.");
    }
    this._port = port;
  }

  onConnect: (connectionId: number) => void;
  onDisconnect: (connectionId: number) => void;
  onMessage: (connectionId: number, message: Uint8Array) => void;

  constructor(
    maxPayloadLength = 2 * 1024 * 1024,
    idleTimeout = 32,
    port = 3000
  ) {
    this._clients = new Map();
    this._ids = new Map();
    this._port = port;

    this._app = uWS
      ./*SSL*/ App({})
      .ws("/", {
        /* Options */
        compression: uWS.SHARED_COMPRESSOR,
        maxPayloadLength,
        idleTimeout,
        /* Handlers */
        open: (ws) => {
          const connectionId = this.server.nextConnectionId;
          this._clients.set(connectionId, ws);
          this._ids.set(ws, connectionId);
          this.onConnect(connectionId);
        },
        message: (ws, message, isBinary) => {
          const connectionId = this._ids.get(ws);
          if (connectionId === undefined)
            throw new Error("Message from unknown connection?");

          if (!isBinary) {
            console.warn(
              "Received non-binary message from client " + connectionId
            );
            return;
          }

          this.onMessage(connectionId, new Uint8Array(message));
        },
        drain: (ws) => {
          // #TODO: HANDLE BACKPRESSURE
          console.log("WebSocket backpressure: " + ws.getBufferedAmount());
        },
        close: (ws, code) => {
          const connectionId = this._ids.get(ws);
          if (connectionId === undefined)
            throw new Error("Closing unknown connection?");

          this.onDisconnect(connectionId);
          this._clients.delete(connectionId);
          this._ids.delete(ws);
        },
      })
      .any("/*", (res, req) => {
        res.end("[connect.js] Server, Transport: uWebSockets.js");
      });
  }

  send(connectionId: number, message: Uint8Array): void {
    this._clients.get(connectionId)?.send(message, true);
  }

  close(connectionId: number): void {
    this._clients.get(connectionId)?.close();
  }

  start(): void {
    this._app.listen(this._port, (token: uWS.us_listen_socket) => {
      if (!token) {
        throw new Error("Failed to listen to port " + this._port);
      }
      this._listen_socket = token;
    });
  }

  stop(): void {
    uWS.us_listen_socket_close(this._listen_socket);
    this._listen_socket = null;
    this._clients.clear();
    this._ids.clear();
  }

  getConnections(): number[] {
    return Array.from(this._clients.keys());
  }
}


src/Server.ts
import { PakettoInstance, paketto } from "paketto";
import { ClientStatus } from "./Enums/ClientStatus";
import { Protocol } from "./Enums/Protocol";
import { SocketClient } from "./SocketClient";
import { ITransport } from "./Transports/ITransport";
import { WebSocketTransport } from "./Transports/WebSocket/WebSocketTransport";

export class Server {
  /**
   * @internal
   */
  private _lastConnectionId = 1;

  /**
   * @internal
   */
  private readonly _buffer: Uint8Array;

  /**
   * @internal
   */
  private _paketto: PakettoInstance;

  get paketto() {
    return this._paketto;
  }

  /**
   * Gets the next connection ID.
   */
  get nextConnectionId() {
    return this._lastConnectionId++;
  }

  get buffer() {
    return this._buffer;
  }

  /**
   * @internal
   */
  private readonly _transports: Set<ITransport> = new Set();
  /**
   * @internal
   */
  private readonly _clients: Map<number, SocketClient> = new Map();

  /**
   * Returns a list of transports. This is a copy of the internal list, so
   * modifying it will not affect the server.
   */
  get transports() {
    return new Set(this._transports);
  }

  /**
   * Returns a list of clients. This is a copy of the internal list, so
   * modifying it will not affect the server.
   */
  get clients() {
    return new Map(this._clients);
  }

  /**
   * Called when a client connects to the server.
   * @param client The client that connected.
   * @event
   */
  onConnect?: (client: SocketClient) => void = () => {};

  /**
   * Creates a new server.
   * @param pakettoInstance The paketto instance to use. Defaults to a new
   * instance.
   * @param preallocateBytes The number of bytes to preallocate for the
   * packet buffer. Defaults to 1MB (1024 * 1024).
   * @param transports The transports to use. If none are provided, a
   * {@link WebSocketTransport} with port 3000 will be used.
   */
  constructor(
    pakettoInstance = paketto(),
    preallocateBytes = 1024 * 1024,
    ...transports: ITransport[]
  ) {
    this._paketto = pakettoInstance;
    this._buffer = new Uint8Array(preallocateBytes);
    if (transports.length === 0) {
      const transport = new WebSocketTransport();
      transport.server = this;
      this._transports.add(transport);
    } else {
      transports.forEach((transport) => {
        transport.server = this;
        this._transports.add(transport);
      });
    }
  }

  /**
   * Starts the server.
   */
  start() {
    this._transports.forEach((transport) => {
      transport.onConnect = (connectionId) => {
        this._onConnect(connectionId, transport);
      };
      transport.onDisconnect = this._onDisconnect.bind(this);
      transport.onMessage = this._onMessage.bind(this);
      transport.start();
    });
  }

  /**
   * Stops the server.
   */
  stop() {
    this._transports.forEach((transport) => {
      transport.stop();
    });
  }

  /**
   * Broadcasts a message to all connected clients.
   * @param event The event identifier.
   * @param data The data to send.
   */
  broadcast(event: unknown, data: unknown) {
    this._clients.forEach((client) => {
      client.send(event, data);
    });
  }

  /**
   * @internal
   */
  private _onConnect(connectionId: number, transport: ITransport) {
    const client = new SocketClient(connectionId, transport, this);
    this._clients.set(connectionId, client);

    client.status = ClientStatus.READY;
    this.onConnect(client);
  }

  /**
   * @internal
   */
  private _onDisconnect(connectionId: number) {
    const client = this._clients.get(connectionId);
    if (!client) throw new Error("Client disconnected but was not connected.");
    client.status = ClientStatus.CLOSED;
    client.onClose();
    this._clients.delete(connectionId);
  }

  /**
   * @internal
   */
  private _onMessage(connectionId: number, message: Uint8Array) {
    const client = this._clients.get(connectionId);
    if (!client) throw new Error("Message received from unknown client.");

    if (message.length === 0) {
      console.warn("Received empty packet.");
      client.close();
      return;
    }

    const packetId = <Protocol>message[0];
    let offset = 1;

    switch (packetId) {
      case Protocol.SC_Message:
        if (client.status !== ClientStatus.READY) {
          console.warn("Received message from client that is not ready.");
          client.close();
          break;
        }
        if (message.length < 2) {
          console.warn("Received message is too short.");
          client.close();
          break;
        }

        // Key
        const deserializedKey = this._paketto.deserialize(message, offset);
        offset = deserializedKey.offset;

        if (message.length < offset + 1) {
          console.warn("Received message is too short.");
          client.close();
          break;
        }

        // Value
        const deserializedValue = this._paketto.deserialize(message, offset);
        offset = deserializedValue.offset;

        // Call the client's onMessage handler
        client.onMessage(
          deserializedKey.value as string,
          deserializedValue.value as any
        );

        break;
      default:
        console.warn(`Received unknown packet ID: ${packetId}`);
        client.close();
        break;
    }
  }
}
